### DISCUSSION TITLE
can pylance ignore error: 'Type of "[some-third part code]" is partially unknown' from third-party libraries?

### DISCUSSION THREAD

**Discussion by raceychan** [microsoft/pyright#6575] [Q&A] [answered]
when we import a third-party library, it is possible that the functions/classes we use from the libary isn't typed completedly, and in this situation, pylance would raise partially unknown error, but it is not really what we can fix immediatedly on our own, is there something configrable on the pylance side to ignore such error?

![image](https://github.com/microsoft/pyright/assets/75417963/acccefd7-6064-44a6-a603-50179e81aa94)

I know that we can always add # type: ignore to supress the error, but i still hope pylance would notice me something is wrong, such as unmatch of known types.

**Comment by erictraut** [comment, marked as answer]
By default, pyright (the type checker upon which pylance is built) does not report the use of `Unknown` types. It does this only if you've enabled "strict mode". These strict-mode checks are designed for people who want to know about any potential "holes" in their type checking. If you're importing types from libraries that have incomplete types, you will naturally have holes in your type checking, so using these strict mode checks doesn't make sense in this case. Your options are: 1) disable the `reportUnknown...` family of diagnostic checks (you can manually enable other strict-mode checks if you'd like), or 2) augment the type information for the libraries you're consuming via type stubs so the type information is complete.

**Reply by raceychan** [reply]
is it possible to combine 'disable the reportUnknown... family of diagnostic checks ' and python.analysis.include together, so that i can put third-party libraries in python.analysis.include to only disable the reportUnknown error in third-part library?

**Reply by erictraut** [reply]
No, pyright doesn't offer such an option. And an option like this doesn't make sense if you think about how types work in a static type checker.

First, it's important to understand that pyright never reports errors in third-party libraries, so `reportUnknown...` doesn't affect third-party code. These checks generate errors in _your own code_ when you're using a type that is unknown or partially unknown and relying on that type for type validation.

Second, a "type" is not the same as a "class". A class is one example of a type, but types are much more complex. They include functions, overloads, unions, specialized generics, type variables. Types are evaluated and transformed throughout the type checking process (e.g. type guards, code flow joins, assignments, the application of decorators, generic function calls, etc.), so the type of a local symbol or an expression at a specific point in the code flow graph often combines type information from many sources. For example, it might be a union of types that includes a generic class that is parameterized by another union that includes a class that was imported from a particular library. There's no way to say "this type came from this library" because types are composed of other types.

The `reportUnknown...` family of checks tell you when you're using an expression whose evaluated type (regardless of how it was composed) is partially unknown.

**Reply by raceychan** [reply]
understood,  types from different library are just type in the eye of pyright and there is no way for pyright to backtracing its origin. thanks for you explaination
